File: pe-core/src/main/java/org/saar/core/model/data/DataWriter.java
package org.saar.core.model.data;

import org.joml.*;

import java.util.ArrayList;
import java.util.List;

public class DataWriter {

    private final List<Integer> data = new ArrayList<>();

    public void write(float value) {
        this.getData().add(Float.floatToIntBits(value));
    }

    public void write(Vector2fc value) {
        this.write(value.x());
        this.write(value.y());
    }

    public void write(Vector3fc value) {
        this.write(value.x());
        this.write(value.y());
        this.write(value.z());
    }

    public void write(Vector4fc value) {
        this.write(value.x());
        this.write(value.y());
        this.write(value.z());
        this.write(value.w());
    }

    public void write(Matrix4fc value) {
        final float[] array = value.get(new float[16]);
        for (float v : array) {
            this.write(v);
        }
    }

    public void write(int value) {
        this.getData().add(value);
    }

    public void write(Vector2ic value) {
        this.write(value.x());
        this.write(value.y());
    }

    public void write(Vector3ic value) {
        this.write(value.x());
        this.write(value.y());
        this.write(value.z());
    }

    public void write(Vector4ic value) {
        this.write(value.x());
        this.write(value.y());
        this.write(value.z());
        this.write(value.w());
    }

    public List<Integer> getData() {
        return this.data;
    }

    public int[] getDataArray() {
        final int size = this.data.size();
        final int[] data = new int[size];
        for (int i = 0; i < size; i++) {
            data[i] = this.data.get(i);
        }
        return data;
        /*return this.data.stream().mapToInt(
                value -> value).toArray();*/
    }

}

File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/util/DataWriter.kt
package org.saar.lwjgl.util

import org.joml.*

interface DataWriter {

    fun writeByte(index: Int, value: Byte)
    fun writeByte(value: Byte)

    fun writeShort(index: Int, value: Short)
    fun writeShort(value: Short)

    fun writeChar(index: Int, value: Char)
    fun writeChar(value: Char)


    fun writeInt(index: Int, value: Int)
    fun writeInt(value: Int)

    fun writeFloat(index: Int, value: Float)
    fun writeFloat(value: Float)

    fun writeLong(index: Int, value: Long)
    fun writeLong(value: Long)

    fun writeDouble(index: Int, value: Double)
    fun writeDouble(value: Double)

    fun write2f(index: Int, value: Vector2fc)
    fun write2f(value: Vector2fc)

    fun write3f(index: Int, value: Vector3fc)
    fun write3f(value: Vector3fc)

    fun write4f(index: Int, value: Vector4fc)
    fun write4f(value: Vector4fc)

    fun write4x4f(index: Int, value: Matrix4fc)
    fun write4x4f(value: Matrix4fc)

    fun write2i(index: Int, value: Vector2ic)
    fun write2i(value: Vector2ic)

    fun write3i(index: Int, value: Vector3ic)
    fun write3i(value: Vector3ic)

    fun write4i(index: Int, value: Vector4ic)
    fun write4i(value: Vector4ic)
}